### PR TITLE
Add error message for negative argument of sqrt intrinsic function

### PR DESCRIPTION
**Pull request by parth121101**
fixes: #3867

### CODE REVIEW

File: src/libasr/pass/intrinsic_functions.h [hunk 1]
```diff
@@ -2075,11 +2075,17 @@ namespace Dim {
...
     static ASR::expr_t *eval_Sqrt(Allocator &al, const Location &loc,
-            ASR::ttype_t* arg_type, Vec<ASR::expr_t*> &args, diag::Diagnostics& /*diag*/) {
+            ASR::ttype_t* arg_type, Vec<ASR::expr_t*> &args, diag::Diagnostics& diag) {
         ASRUtils::ASRBuilder b(al, loc);
         if (is_real(*arg_type)) {
             double val = ASR::down_cast<ASR::RealConstant_t>(expr_value(args[0]))->m_r;
-            return b.f(std::sqrt(val), arg_type);
+            if (val < 0.0) {
+            append_error(diag, "Argument of SQRT has a negative value", loc);
+            return nullptr;
```
Comment (HarshitaKalani) [inline src/libasr/pass/intrinsic_functions.h:2084]:
```suggestion
            if (val < 0.0) {
                append_error(diag, "Argument of SQRT has a negative value", loc);
                return nullptr;
```

**Review by HarshitaKalani — CHANGES_REQUESTED** [review]
Can you also add an error test to check if it works perfectly?

File: src/libasr/pass/intrinsic_functions.h [hunk 2]
```diff
@@ -2075,11 +2075,17 @@ namespace Dim {
...
+            }
+            else{
+                return b.f(std::sqrt(val), arg_type);
+            }
```
Comment (Shaikh-Ubaid) [inline src/libasr/pass/intrinsic_functions.h:2088]: The else here seems redundant.

```suggestion
            return b.f(std::sqrt(val), arg_type);
```

Comment (parth121101) [reply src/libasr/pass/intrinsic_functions.h]: @Shaikh-Ubaid I have added the error message for the negative argument. But still, the function is returning nan values along with the error message.

```
cat a.f90
program sqrt_neg
    print *, sqrt(-1.0)
end program

lfortran a.f90
semantic error: Argument of SQRT has a negative value
 --> a.f90:2:14
  |
2 |     print *, sqrt(-1.0)
  |              ^^^^^^^^^^ 

Note: Please report unclear, confusing or incorrect messages as bugs at
https://github.com/lfortran/lfortran/issues.
         -nan
```

Comment (Shaikh-Ubaid) [reply src/libasr/pass/intrinsic_functions.h]: https://github.com/lfortran/lfortran/blob/f6d9f0f9504939cfa143e949811f01265c7ef654/src/libasr/intrinsic_func_registry_util_gen.py#L704-L705

It looks like we need to add the following check in the above location.

```console
if (diagnostics.has_error()) {
    return nullptr
}

Comment (Shaikh-Ubaid) [reply src/libasr/pass/intrinsic_functions.h]: We might need to add the same check after [this](https://github.com/lfortran/lfortran/blob/f6d9f0f9504939cfa143e949811f01265c7ef654/src/libasr/intrinsic_func_registry_util_gen.py#L694) line as well.

File: src/libasr/pass/intrinsic_functions.h [hunk 3]
```diff
@@ -2075,10 +2075,14 @@ namespace Dim {
...
 
     static ASR::expr_t *eval_Sqrt(Allocator &al, const Location &loc,
-            ASR::ttype_t* arg_type, Vec<ASR::expr_t*> &args, diag::Diagnostics& /*diag*/) {
+            ASR::ttype_t* arg_type, Vec<ASR::expr_t*> &args, diag::Diagnostics& diag) {
         ASRUtils::ASRBuilder b(al, loc);
         if (is_real(*arg_type)) {
             double val = ASR::down_cast<ASR::RealConstant_t>(expr_value(args[0]))->m_r;
+            if (val < 0.0) {
+                append_error(diag, "Argument of SQRT has a negative value", loc);
+            return nullptr;
```
Comment (Shaikh-Ubaid) [inline src/libasr/pass/intrinsic_functions.h:2084]:
```suggestion
                return nullptr;
```

File: src/libasr/pass/intrinsic_functions.h [hunk 4]
```diff
@@ -2075,10 +2075,14 @@ namespace Dim {
 namespace Sqrt {
 
     static ASR::expr_t *eval_Sqrt(Allocator &al, const Location &loc,
-            ASR::ttype_t* arg_type, Vec<ASR::expr_t*> &args, diag::Diagnostics& /*diag*/) {
+            ASR::ttype_t* arg_type, Vec<ASR::expr_t*> &args, diag::Diagnostics& diag) {
         ASRUtils::ASRBuilder b(al, loc);
         if (is_real(*arg_type)) {
             double val = ASR::down_cast<ASR::RealConstant_t>(expr_value(args[0]))->m_r;
```
Comment (Shaikh-Ubaid) [inline src/libasr/pass/intrinsic_functions.h:2081]: Unrelated to your PR:
```suggestion
            double val = ASR::down_cast<ASR::RealConstant_t>(args[0])->m_r;
```
I think `expr_value()` is already computed and passed in `args`. So, we need not compute it here again. We should update this for other `eval_*()` functions as well (But in a separate PR).

File: src/libasr/pass/intrinsic_functions.h [hunk 5]
```diff
@@ -2075,10 +2075,14 @@ namespace Dim {
...
+            if (val < 0.0) {
+                append_error(diag, "Argument of SQRT has a negative value", loc);
```
Comment (certik) [inline src/libasr/pass/intrinsic_functions.h:2083]: I prefer using lowercase:
```suggestion
                append_error(diag, "Argument of `sqrt` has a negative value", loc);
```
The uppercase reminds me of old F77 style, while lowercase is how most modern languages are used.

File: tests/errors/sqrt_neg.f90 [hunk 6]
```diff
@@ -0,0 +1,3 @@
+program sqrt_neg
+    print *, sqrt(-1.0)
```
Comment (certik) [inline tests/errors/sqrt_neg.f90:2]: Can you register a test for this and add reference results?

File: src/libasr/intrinsic_func_registry_util_gen.py [hunk 7]
```diff
@@ -702,6 +705,9 @@ def add_create_func_return_src(func_name):
         for _i in range(no_of_args):
             src += indent * 3 + f"args_values.push_back(al, expr_value(m_args[{_i}]));\n"
         src += indent * 3 +     f"m_value = eval_{func_name}(al, loc, return_type, args_values, diag);\n"
+        src += indent * 2 + "if (diag.has_error()) {\n"
+        src += indent * 3 +     f"return nullptr;\n"
+        src += indent * 2 + "}\n"
```
Comment (Shaikh-Ubaid) [inline src/libasr/intrinsic_func_registry_util_gen.py:710]: It looks like this should be:
```suggestion
        src += indent * 3 +     "if (diag.has_error()) {\n"
        src += indent * 4 +         f"return nullptr;\n"
        src += indent * 3 +     "}\n"
```

**Review by Shaikh-Ubaid — APPROVED** [review]
Shared a minor coment. Rest looks good to me. Thanks for this. Great work!

File: src/libasr/intrinsic_func_registry_util_gen.py (hunk 7 above)
Comment (parth121101) [reply src/libasr/intrinsic_func_registry_util_gen.py]: Done

File: src/libasr/intrinsic_func_registry_util_gen.py [hunk 8]
```diff
@@ -702,6 +705,9 @@ def add_create_func_return_src(func_name):
         for _i in range(no_of_args):
             src += indent * 3 + f"args_values.push_back(al, expr_value(m_args[{_i}]));\n"
         src += indent * 3 +     f"m_value = eval_{func_name}(al, loc, return_type, args_values, diag);\n"
+        src += indent * 3 + "if (diag.has_error()) {\n"
+        src += indent * 4 +     f"return nullptr;\n"
+        src += indent * 3 + "}\n"
```
Comment (Shaikh-Ubaid) [inline src/libasr/intrinsic_func_registry_util_gen.py:710]:
```suggestion
        src += indent * 3 +     "if (diag.has_error()) {\n"
        src += indent * 4 +         f"return nullptr;\n"
        src += indent * 3 +     "}\n"
```

**Review by Shaikh-Ubaid — APPROVED** [review]

**Pull request merged**